File: c#/03-LiskovSubstitutionPrinciple.cs

// Liskov Substitution Principle Violation

class Duck {

    public string Fly() {}

    public string Swim() {}

    public string Cuack() {}
}

class RubberDuck : Duck {

    public string Fly() {
        throw new Exception("Not implemented");
    }

    public string Swim() {
        return "Swim";
    }

    public string Cuack() {
        return "Cuack";
    }
}

class DuckProcesser {

    public void MakeDucksFly(Duck[] ducks) {

        try {

            foreach(Duck duck in ducks) {
                Console.WriteLine(duck.fly());
            }

        } catch (Exception err) {
            Console.WriteLine(err);
        }

    }
}

// Refactored

interface IFly {

    string Fly();
}

interface ISwim {

    string Swim();
}

interface ICuack {

    string Cuack();
}


class Duck : IFly, ISwim, ICuack {

    public string Fly() {}

    public string Swim() {}

    public string Cuack() {}
}

class RubberDuck : IFly, ISwim {

    public string Swim() {
        return "Swim";
    }

    public string Cuack() {
        return "Cuack";
    }
}

class DuckProcesser {

    public void MakeDucksFly(Duck[] ducks) {

        foreach(Duck duck in ducks) {
            Console.WriteLine(duck.fly());
        }

    }
}

File: js/03-liskovSubstitutionPrinciple.js

// Liskov Substitution Principle Violation

class Duck {

    fly() {
        return 'fly';
    }

    swim() {
        return 'swim';
    }

    cuack() {
        return 'cuack';
    }
}

class RubberDuck {

    fly() {
        throw new Error('Not implemented');
    }

    swim() {
        return 'Swim';
    }

    cuack() {
        return 'Cuack';
    }
}

class DuckProcesser {

    makeDucksFly(ducks) {

        try {

            for(let duck of ducks) {
                console.log(duck.fly());
            }

        } catch (err) {
            console.log(err);
        }

    }
}

// Refactored

class Duck {

    fly() {
        return 'fly';
    }

    swim() {
        return 'swim';
    }

    cuack() {
        return 'cuack';
    }
}

class RubberDuck {

    swim() {
        return 'swim';
    }

    cuack() {
        return 'cuack';
    }
}

class DuckProcesser {

    makeDucksFly(ducks) {

        for(let duck of ducks) {
            console.log(duck.fly());
        }
    }
}
